### PR TITLE
Sonarjs 434

### PR DESCRIPTION
**Pull request by vilchik-elena**

### CODE REVIEW

File: javascript-squid/src/main/java/org/sonar/javascript/ast/resolve/type/Utils.java [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02
+ */
+package org.sonar.javascript.ast.resolve.type;
+
+import org.sonar.plugins.javascript.api.tree.expression.ExpressionTree;
+import org.sonar.plugins.javascript.api.tree.expression.IdentifierTree;
+
+public class Utils {
```
Comment (SonarTech) [inline javascript-squid/src/main/java/org/sonar/javascript/ast/resolve/type/Utils.java]: MAJOR Add a private constructor to hide the implicit public one. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1118)

File: javascript-squid/src/main/java/org/sonar/javascript/ast/resolve/type/TypeVisitor.java [hunk 2]
```diff
@@ -165,14 +179,43 @@ public void visitIdentifier(IdentifierTree tree) {
...
+    }
+
+    if (WebAPI.isDocument(tree)){
+      ((IdentifierTreeImpl) tree).addType(ObjectType.WebApiType.DOCUMENT);
+    }
   }
 
   @Override
   public void visitParenthesisedExpression(ParenthesisedExpressionTree tree) {
+    // todo (Lena) : change these lines' order?
```
Comment (SonarTech) [inline javascript-squid/src/main/java/org/sonar/javascript/ast/resolve/type/TypeVisitor.java]: INFO Complete the task associated to this TODO comment. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1135)

File: javascript-squid/src/main/java/org/sonar/javascript/ast/resolve/type/WebAPI.java [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+    return false;
+  }
+
+
+
+  public static boolean isDocument(IdentifierTree tree) {
+    return tree.name().equals(DOCUMENT);
+  }
+
+  public static boolean isElement(ExpressionTree tree) {
```
Comment (SonarTech) [inline javascript-squid/src/main/java/org/sonar/javascript/ast/resolve/type/WebAPI.java]: MAJOR The Cyclomatic Complexity of this method "isElement" is 12 which is greater than 10 authorized. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

File: javascript-squid/src/main/java/org/sonar/javascript/ast/resolve/type/WebAPI.java [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+      "documentElement",
+      "pointerLockElement"
+  );
+
+
+
+  private WebAPI(){
+  }
+
+  public static boolean isWindow(ExpressionTree tree){
```
Comment (SonarTech) [inline javascript-squid/src/main/java/org/sonar/javascript/ast/resolve/type/WebAPI.java]: MAJOR The Cyclomatic Complexity of this method "isWindow" is 19 which is greater than 10 authorized. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)
MAJOR Reduce the number of returns of this method 6, down to the maximum allowed 5. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1142)

File: javascript-squid/src/main/java/org/sonar/javascript/ast/resolve/type/WebAPI.java [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+  }
+
+
+
+  public static boolean isDocument(IdentifierTree tree) {
+    return tree.name().equals(DOCUMENT);
+  }
+
+  public static boolean isElement(ExpressionTree tree) {
+    // todo (Lena): code duplication here
```
Comment (SonarTech) [inline javascript-squid/src/main/java/org/sonar/javascript/ast/resolve/type/WebAPI.java]: INFO Complete the task associated to this TODO comment. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1135)

**Comment by SonarTech** [comment]
SonarQube analysis reported 6 new issues:
* 4 major
* 2 info

**Comment by lindamartin** [comment]
Well done! LGTM!

**Comment by SonarTech** [comment]
SonarQube analysis reported 9 issues:
* MAJOR 8 major
* INFO 1 info

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR Add a 'package-info.java' file to document the 'src/main/java/org/sonar/javascript/checks' package (org.codehaus.sonar-plugins.javascript:javascript-checks:src/main/java/org/sonar/javascript/checks) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1228)
* MAJOR [This method has 111 lines, which is greater than the 100 lines authorized. Split it into smaller methods.](https://github.com/SonarCommunity/sonar-javascript/blob/765b5e0a40204e6cf84f5a761d51f826be10fb3d/javascript-checks/src/main/java/org/sonar/javascript/checks/CheckList.java#L37) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS138)
* MAJOR Add a 'package-info.java' file to document the 'src/main/java/org/sonar/javascript/ast/resolve' package (org.codehaus.sonar-plugins.javascript:javascript-squid:src/main/java/org/sonar/javascript/ast/resolve) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1228)
* INFO [Complete the task associated to this TODO comment.](https://github.com/SonarCommunity/sonar-javascript/blob/765b5e0a40204e6cf84f5a761d51f826be10fb3d/javascript-squid/src/main/java/org/sonar/javascript/ast/resolve/SymbolDeclarationVisitor.java#L157) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1135)
* MAJOR Add a 'package-info.java' file to document the 'src/main/java/org/sonar/javascript/ast/resolve/type' package (org.codehaus.sonar-plugins.javascript:javascript-squid:src/main/java/org/sonar/javascript/ast/resolve/type) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1228)
* MAJOR [Make "inferParameterType" a "static" method.](https://github.com/SonarCommunity/sonar-javascript/blob/765b5e0a40204e6cf84f5a761d51f826be10fb3d/javascript-squid/src/main/java/org/sonar/javascript/ast/resolve/type/TypeVisitor.java#L138) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS2325)
* MAJOR Add a 'package-info.java' file to document the 'src/main/java/org/sonar/javascript/model/internal/expression' package (org.codehaus.sonar-plugins.javascript:javascript-squid:src/main/java/org/sonar/javascript/model/internal/expression) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1228)
* MAJOR Add a 'package-info.java' file to document the 'src/main/java/org/sonar/plugins/javascript/api/symbols' package (org.codehaus.sonar-plugins.javascript:javascript-squid:src/main/java/org/sonar/plugins/javascript/api/symbols) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1228)
* MAJOR Add a 'package-info.java' file to document the 'src/main/java/org/sonar/plugins/javascript/api/tree/expression' package (org.codehaus.sonar-plugins.javascript:javascript-squid:src/main/java/org/sonar/plugins/javascript/api/tree/expression) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1228)

**Pull request closed without merging**